unixfs/src/walk.rs

```diff
@@ -0,0 +1,1040 @@
+                // path items.
+
+                let mut pending = {
+                    let mut pending = self.pending;
+                    for link in links {
+                        pending.push(link?);
+                    }
+                    pending
+                };
+
+                let state = if let Some(next) = pending.pop() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next: Some(next),
+                        pending,
+                    })
+                } else {
+                    State::Last(self.current.unwrap())
+                };
+
+                Ok(ContinuedWalk::Directory(Item::from(state)))
+            }
+            UnixFsType::Raw | UnixFsType::File => {
+                let (bytes, file_size, metadata, step) =
+                    IdleFileVisit::default().start_from_parsed(flat, cache)?;
+
+                let (cid, name, depth) = self.next.expect("validated at new and this method");
+                let file_continues = step.is_some();
+
+                match self.current.as_mut() {
+                    Some(current) => current.as_file(cid, &name, depth, metadata, step, file_size),
+                    _ => {
+                        self.current = Some(InnerEntry::new_root_file(
+                            cid, metadata, &name, step, file_size, depth,
+                        ))
+                    }
+                }
+
+                let next = self.pending.pop();
+                let segment = FileSegment::first(bytes, !file_continues);
+
+                let state = if file_continues || next.is_some() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next,
+                        pending: self.pending,
+                    })
+                } else {
+                    State::Last(self.current.unwrap())
+                };
+
+                Ok(ContinuedWalk::File(segment, Item::from(state)))
+            }
+            UnixFsType::Metadata => Err(Error::UnsupportedType(flat.data.Type.into())),
+            UnixFsType::Symlink => {
+                let contents = match flat.data.Data {
+                    Some(Cow::Borrowed(bytes)) if !bytes.is_empty() => bytes,
+                    None | Some(Cow::Borrowed(_)) => &[][..],
+                    _ => unreachable!("never used into_owned"),
+                };
+
+                let (cid, name, depth) = self.next.expect("continued without next");
+                match self.current.as_mut() {
+                    Some(current) => current.as_symlink(cid, &name, depth, metadata),
+                    _ => {
+                        self.current =
+                            Some(InnerEntry::new_root_symlink(cid, metadata, &name, depth));
+                    }
+                }
+
+                let state = if let Some(next) = self.pending.pop() {
```

Comment:
nit: `if let next @ Some(_) = self.pending.pop()` would allow to just use `next` 3 lines later

Reply:
```suggestion
                let state = if let next @ Some(_) = self.pending.pop() {
```